### PR TITLE
[PIL-732][AV] - RFM Contact Name and Guidance Page

### PR DESCRIPTION
**Pull request by aashik-varahan**

### CODE REVIEW

File: app/cache/SessionData.scala [hunk 1]
```diff
@@ -42,4 +42,8 @@ class SessionData {
     request.session +
       (Pillar2SessionKeys.updateMneOrDomestic -> value)
 
+  def corporatePosition(value: String)(implicit request: Request[AnyContent]): Session =
```
Comment (alikouhestani) [inline app/cache/SessionData.scala:45]: We no longer store data on Session. You need to store any non-authenticated journey data on sessionRepository

File: app/controllers/rfm/CorporatePositionController.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+              _              <- userAnswersConnectors.save(updatedAnswers.id, Json.toJson(updatedAnswers.data))
+            } yield Redirect(controllers.routes.UnderConstructionController.onPageLoad)
+              .withSession((sessionData.corporatePosition(value.toString)))
+
+          case value @ CorporatePosition.NewNfm =>
+            for {
+              updatedAnswers <- Future.fromTry(request.userAnswers.set(rfmCorporatePositionPage, value))
+              _              <- userAnswersConnectors.save(updatedAnswers.id, Json.toJson(updatedAnswers.data))
+            } yield Redirect(controllers.routes.UnderConstructionController.onPageLoad)
+              .withSession((sessionData.corporatePosition(value.toString)))
```
Comment (alikouhestani) [inline app/controllers/rfm/CorporatePositionController.scala:83]: again, please refrain from storing data on session as this data is public across HMRC and its frowned upon to store data like this

**Review by alikouhestani — CHANGES_REQUESTED** [review]

**Review by alikouhestani — APPROVED** [review]

File: app/controllers/rfm/UpeStartPageRegistrationController.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.rfm.UpeStartPageRegistrationView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class UpeStartPageRegistrationController @Inject() (
+  rfmIdentify:              RfmIdentifierAction,
+  val controllerComponents: MessagesControllerComponents,
+  view:                     UpeStartPageRegistrationView
```
Comment (alikouhestani) [inline app/controllers/rfm/UpeStartPageRegistrationController.scala:35]: Is the naming correct for this one? Why is it called a "Upe" (Ultimate Parent Entity) if this a part of the RFM (Replace Filing Member) Journey

File: app/forms/RfmUpeNameRegistrationFormProvider.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package forms
+
+import forms.mappings.Mappings
+import mapping.Constants
+import play.api.data.Form
+
+import javax.inject.Inject
+
```
Comment (alikouhestani) [inline app/forms/RfmUpeNameRegistrationFormProvider.scala:24]: again, is the naming right?

File: app/pages/rfmUpeNameRegistrationPage.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pages
+
+import play.api.libs.json.JsPath
```
Comment (alikouhestani) [inline app/pages/rfmUpeNameRegistrationPage.scala:19]: see above

File: app/views/rfm/UpeNameRegistrationView.scala.html [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+layout: templates.Layout,
+formHelper: FormWithCSRF,
+sectionHeader: sectionHeader,
+govukErrorSummary: GovukErrorSummary,
+govukInput: GovukInput,
+govukButton: GovukButton
+)
+
+@(form: Form[_], mode: Mode)(implicit request: Request[_], appConfig: FrontendAppConfig, messages: Messages)
+
```
Comment (alikouhestani) [inline app/views/rfm/UpeNameRegistrationView.scala.html:32]: see above

File: app/views/rfm/UpeStartPageRegistrationView.scala.html [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@import models.NormalMode
+@import config.FrontendAppConfig
+@import views.html.components.gds._
+
+@this(
```
Comment (alikouhestani) [inline app/views/rfm/UpeStartPageRegistrationView.scala.html:21]: see above

File: test/controllers/rfm/CorporatePositionControllerSpec.scala [hunk 8]
```diff
@@ -107,7 +107,7 @@ class CorporatePositionControllerSpec extends SpecBase {
       }
     }
 
-    "must redirect to the under construction page when valid data is submitted with UPE" in {
+    "must redirect to the UPE registration start page when valid data is submitted with UPE" in {
 
       val application = applicationBuilder(userAnswers = None)
         .overrides(
```
Comment (alikouhestani) [inline test/controllers/rfm/CorporatePositionControllerSpec.scala:113]: .configure(
          Seq(
            "features.rfmAccessEnabled" -> true
          ): _*
        )
        why do we need this if the toggle is set to true in local anyway?

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+import pages.rfmUpeNameRegistrationPage
+import play.api.inject
+import play.api.libs.json.Json
+import play.api.test.FakeRequest
+import play.api.test.Helpers._
+import views.html.rfm.UpeNameRegistrationView
+
+import scala.concurrent.Future
+
+class UpeNameRegistrationControllerSpec extends SpecBase {
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:34]: Same issue with the naming convention

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      }
+    }
+
+    "must return OK and populate the view correctly when the question has been previously answered" in {
+      val userAnswers = emptyUserAnswers.setOrException(rfmUpeNameRegistrationPage, "name")
+
+      val application = applicationBuilder(userAnswers = Some(userAnswers))
+        .configure(
+          Seq(
+            "features.rfmAccessEnabled" -> true
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:68]: again this toggle is set to true in local so this is redundant

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      }
+    }
+
+    "must redirect to the under construction page when rfm feature is set to false" in {
+
+      val ua = emptyUserAnswers
+      val application = applicationBuilder(userAnswers = Some(ua))
+        .configure(
+          Seq(
+            "features.rfmAccessEnabled" -> false
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:93]: see above

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+    "must redirect to the under construction page when valid data is submitted" in {
+
+      val application = applicationBuilder(userAnswers = None)
+        .overrides(
+          inject.bind[UserAnswersConnectors].toInstance(mockUserAnswersConnectors)
+        )
+        .configure(
+          Seq(
+            "features.rfmAccessEnabled" -> true
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:117]: see above

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        redirectLocation(result).value mustEqual controllers.routes.UnderConstructionController.onPageLoad.url
+      }
+    }
+
+    "must return a Bad Request and errors when invalid data is submitted" in {
+
+      val application = applicationBuilder(userAnswers = None)
+        .configure(
+          Seq(
+            "features.rfmAccessEnabled" -> true
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:140]: see above

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  val formProvider = new RfmUpeNameRegistrationFormProvider()
+
+  "RFM UPE Name Registration controller" when {
+
+    "must return OK and the correct view for a GET" in {
+      val ua = emptyUserAnswers
+      val application = applicationBuilder(userAnswers = Some(ua))
+        .configure(
+          Seq(
+            "features.rfmAccessEnabled" -> true
```
Comment (alikouhestani) [inline test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:45]: see above

**Review by alikouhestani — CHANGES_REQUESTED** [review]

File: app/controllers/rfm/UpeStartPageRegistrationController.scala (hunk 3 above)
Comment (aashik-varahan) [reply app/controllers/rfm/UpeStartPageRegistrationController.scala:35]: @alikouhestani - As previously mentioned in other tickets, following FM validation, the UPE can assume the NFM role, or a completely new NFM entity can replace the existing one. This ticket pertains to selecting the UPE radio button as part of PIL-715.

We may need a broader discussion to determine naming conventions, but I presume it aligns with the UPE and NFM journey in Reg and Sub.

File: app/forms/RfmUpeNameRegistrationFormProvider.scala (hunk 4 above)
Comment (aashik-varahan) [reply app/forms/RfmUpeNameRegistrationFormProvider.scala:24]: Check reply above.

File: app/pages/rfmUpeNameRegistrationPage.scala (hunk 5 above)
Comment (aashik-varahan) [reply app/pages/rfmUpeNameRegistrationPage.scala:19]: Check reply above.

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala (hunk 11 above)
Comment (aashik-varahan) [reply test/controllers/rfm/UpeNameRegistrationControllerSpec.scala:83]: Redundant code for toggle config is now removed.

File: test/controllers/rfm/UpeNameRegistrationControllerSpec.scala (hunk 10 above)
Comment (aashik-varahan) [reply test/controllers/rfm/UpeNameRegistrationControllerSpec.scala]: Redundant code for toggle config is now removed.

File: app/controllers/rfm/UpeStartPageRegistrationController.scala (hunk 3 above)
Comment (alikouhestani) [reply app/controllers/rfm/UpeStartPageRegistrationController.scala:35]: @aashik-varahan 
They can end up on this start page whether if they are a new fm replacing the old one or an existing upe (feel free to check the Mural). Therefore, calling it "UpeStartPageRegistration" indicates that this journey is somehow related to the Ultimate parent which is not.

Comment (aashik-varahan) [reply app/controllers/rfm/UpeStartPageRegistrationController.scala]: @alikouhestani - Renamed the controller class now. It is discussed with @AKKUM that we will follow adding a Prefix 'Rfm' to the name of the files within RFM journey going forward.

**Review by alikouhestani — APPROVED** [review]

**Pull request merged**